### PR TITLE
Add guard based on ReactElement

### PR DESCRIPTION
**Pull request by sesh92**
Allow using tagged state via Guard based on ReactElement

```
<Guard
  uninit={<UninitComponent />}
  ready={<ReadyComponent />}
/>
```

### CODE REVIEW

**Review by MOZGIII — CHANGES_REQUESTED** [review]
I've thought about the design, and figured it wouldn't be fair to prioritise the component guard  or element guard as a default now, since we have to options. So, let's make it so there are two entities: ComponentGuard and ElementGuard. We shall build all the accompanying machinery atround each if them too.

File: packages/react/src/ComponentGuard.tsx [hunk 1]
```diff
@@ -1,7 +1,7 @@
 import React from "react";
 import { AnyTaggedState, StateProps } from "@tagged-state/core";
 
-export type StateComponentProps<
+export type StateComponentGuardProps<
```
Comment (MOZGIII) [inline packages/react/src/ComponentGuard.tsx:4]:
```suggestion
export type StateComponentProps<
```

Comment (sesh92) [reply packages/react/src/ComponentGuard.tsx]: https://github.com/MOZGIII/tagged-state/pull/2/commits/93a787de5e0b59733f11022acc744bfdd6a20abd

File: packages/react/src/createComponentGuard.tsx [hunk 2]
```diff
@@ -1,20 +1,20 @@
...
+import Guard, { StateComponentProps } from "./ComponentGuard";
 
-export type ContextGuardProps<
+export type ContextComponentGuardProps<
   State extends AnyTaggedState<Tag>,
   Tag extends string
 > = StateComponentProps<State, Tag>;
 
-function createContextGuard<
+function createComponentGuard<
```
Comment (MOZGIII) [inline packages/react/src/createComponentGuard.tsx:10]:
```suggestion
function createContextComponentGuard <
```

File: packages/react/src/createElementGuard.tsx [hunk 3]
```diff
@@ -0,0 +1,26 @@
+import React, { useContext } from "react";
+import { AnyTaggedState } from "@tagged-state/core";
+import ElementGuard, { StateElementGuardProps } from "./ElementGuard";
+
+export type ElementGuardProps<
+  State extends AnyTaggedState<Tag>,
+  Tag extends string
+> = StateElementGuardProps<State, Tag>;
+
+function createElementGuard<
```
Comment (MOZGIII) [inline packages/react/src/createElementGuard.tsx:10]:
```suggestion
function createContextElementGuard<
```

File: packages/react/src/createElementGuard.tsx [hunk 4]
```diff
@@ -0,0 +1,26 @@
+import React, { useContext } from "react";
+import { AnyTaggedState } from "@tagged-state/core";
+import ElementGuard, { StateElementGuardProps } from "./ElementGuard";
+
+export type ElementGuardProps<
```
Comment (MOZGIII) [inline packages/react/src/createElementGuard.tsx:5]:
```suggestion
export type ContextElementGuardProps<
```

File: packages/react/src/createElementGuard.tsx [hunk 5]
```diff
@@ -0,0 +1,26 @@
...
+  Tag extends string
+> = StateElementGuardProps<State, Tag>;
+
+function createElementGuard<
+  Context extends React.Context<State>,
+  State extends AnyTaggedState<Tag> = React.ContextType<Context>,
+  // eslint-disable-next-line @typescript-eslint/no-explicit-any
+  Tag extends string = any
+>(context: Context): React.FC<ElementGuardProps<State, Tag>> {
+  function ContextGuard(
```
Comment (MOZGIII) [inline packages/react/src/createElementGuard.tsx:16]:
```suggestion
  function ContextElementGuard(
```

File: packages/react/src/ElementGuard.tsx [hunk 6]
```diff
@@ -0,0 +1,27 @@
+import React from "react";
+import { AnyTaggedState, StateProps } from "@tagged-state/core";
+
+export type StateElementGuardProps<
```
Comment (MOZGIII) [inline packages/react/src/ElementGuard.tsx:4]:
```suggestion
export type StateElementProps<
```

**Review by MOZGIII — APPROVED** [review]

**Pull request merged**